core/templates/domain/story/story-update.service.ts

```diff
@@ -19,664 +19,693 @@
-      STORY_NODE_PROPERTY_EXPLORATION_ID,
-      STORY_NODE_PROPERTY_OUTLINE, STORY_NODE_PROPERTY_PREREQUISITE_SKILL_IDS,
-      STORY_NODE_PROPERTY_THUMBNAIL_BG_COLOR,
-      STORY_NODE_PROPERTY_THUMBNAIL_FILENAME, STORY_NODE_PROPERTY_TITLE,
-      STORY_PROPERTY_DESCRIPTION, STORY_PROPERTY_LANGUAGE_CODE,
-      STORY_PROPERTY_META_TAG_CONTENT,
-      STORY_PROPERTY_NOTES, STORY_PROPERTY_THUMBNAIL_BG_COLOR,
-      STORY_PROPERTY_THUMBNAIL_FILENAME, STORY_PROPERTY_TITLE,
-      STORY_PROPERTY_URL_FRAGMENT) {
-    // Creates a change using an apply function, reverse function, a change
-    // command and related parameters. The change is applied to a given
-    // story.
-    var _applyChange = function(story, command, params, apply, reverse) {
-      var changeDict = angular.copy(params);
-      changeDict.cmd = command;
-      var changeObj = new Change(changeDict, apply, reverse);
-      try {
-        UndoRedoService.applyChange(changeObj, story);
-      } catch (err) {
-        AlertsService.addWarning(err.message);
-        throw err;
-      }
-    };
-
-    var _getParameterFromChangeDict = function(changeDict, paramName) {
-      return changeDict[paramName];
-    };
-
-    var _getNodeIdFromChangeDict = function(changeDict) {
-      return _getParameterFromChangeDict(changeDict, 'node_id');
-    };
-
-    var _getStoryNode = function(storyContents, nodeId) {
-      var storyNodeIndex = storyContents.getNodeIndex(nodeId);
-      if (storyNodeIndex === -1) {
-        throw new Error('The given node doesn\'t exist');
-      }
-      return storyContents.getNodes()[storyNodeIndex];
-    };
-
-    // Applies a story property change, specifically. See _applyChange()
-    // for details on the other behavior of this function.
-    var _applyStoryPropertyChange = function(
-        story, propertyName, oldValue, newValue, apply, reverse) {
-      _applyChange(story, CMD_UPDATE_STORY_PROPERTY, {
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { AlertsService } from 'services/alerts.service';
+import { BackendChangeObject, Change } from 'domain/editor/undo_redo/change.model';
+import cloneDeep from 'lodash/cloneDeep';
+import { UndoRedoService } from 'domain/editor/undo_redo/undo-redo.service.ts';
+import { StoryChange } from 'domain/editor/undo_redo/change.model';
+import { StoryDomainConstants } from 'domain/story/story-domain.constants.ts';
+import { StoryEditorStateService } from 'pages/story-editor-page/services/story-editor-state.service.ts';
+import { Story } from 'domain/story/StoryObjectFactory.ts';
+import { StoryContents } from 'domain/story/StoryContentsObjectFactory.ts';
+import { StoryNode } from './story-node.model';
+
+type StoryUpdateApply = (storyChange: StoryChange, story: Story) => void;
+type StoryUpdateReverse = (storyChange : StoryChange, story: Story) => void;
+
+type Params = {
```

Comment:
Can we use interface here instead of type? Usually, types end up generating a lot of code in .d.ts files (not that it is concerning to us but still, if possible, change to an interface.

Reply:
I have changed params to interface Since most of the types are imported from change.model, They are explicitly using type, I'm getting errors while converting them into an interface. So I'm leaving the rest of the types as it is.